Форматирование и оптимизация кода.

Kirchik opened the pull request:
Заменил массивы песен, групп, популярности на структуру данных.
Добавил вторую структуру для песен и групп в топе.
Добавил процедуру вывода информации о песнях, порядка их воспроизведения.
Добавил процедуры для выполнения повторяющихся участков кода в программе.
Заменил все названия переменных на новые.

Kirchik committed:
Структура данных.

Заменил массивы песен, групп, популярности на структуру данных.
В 4 - 40 строках изменил названия переменных
В 57 - 70 строках изменил способ вывода текста на экран для удобочитаемости кода.

Kirchik committed:
Добавил функцию вывода информации о песнях.

Заменил все сообщения о песнях функцией.
Заменил переменные на новые до строки 131 (до case 4).

Kirchik committed:
Добавил процедуру для сортировки песен.

Добавил процедуру для вывода порядка воспроизведения композиций на экран.
Изменил в некоторых местах названия переменных. (до строки 234 (case11))

Kirchik committed:
Заменил повторяющиеся участки кода функциями.

Заменил повторяющиеся участки кода функциями во всех case.
Изменил основные названия переменных на новые (структуры, константы) во всех case.

Kirchik committed:
Добавил вторую структуру, использовал std::string.

Также воспользовался методом .c_str() для конвертации строк.

Kirchik committed:
Final

Добавил комментарии.
Отформатировал текст.
Изменил имена всех переменных.

g4rfan commented on main.cpp:
```diff
@@ -1,462 +1,206 @@
+	std::string track;
+	std::string group;
+	int popularity;
+}allSongs[AMOUNT_OF_SONGS];
```
Конечно мелочь, но всё же это важно между } и allSongs должен быть пробел. Аккуратность и стиль нужно выдерживать, поверь в реальной жизни к этому очееееень сильно прикапываются.

g4rfan commented on main.cpp:
```diff
@@ -1,462 +1,206 @@
+	std::string topSongs;
+	std::string topGroups;
+	int popularityIndex;
+}topList[AMOUNT_OF_SONGS];
```
Тоже пробела не хватает

g4rfan commented on main.cpp:
```diff
@@ -1,462 +1,206 @@
+
+	bool continueWork = true;
+
+    while (continueWork) { 
```
Тут немного поехало форматирование

Kirchik committed:
Fix. Форматирование, отступы.

Поправил форматирование по while.
Между описанием и объявлением структур сделал отступ.

g4rfan commented on main.cpp:
```diff
@@ -466,4 +213,126 @@ const char* vremenoeNazvanyePesny = pesny_VTope[i];
\ No newline at end of file
+}
+
+void printInformation(int currSong, int all) { // процедура вывода информации о текущей песне, следующей и предыдущей
```
Итак, вроде всё хорошо ты сделал функцию которая выводит информацию о воспроизводимом треке. Она принимает номер текущей песни и все. Что такое "все"?  Эта переменная звучит как будто она типа bool. например processSongs(bool all) здесь бы она означала обработать ли все песни. Вместо all лучше использовать общепринятые size или ещё лучше count, или даже limit но не all. currSong можно было бы написать currentSong всего на 3 символа длиннее, но читается лучше или даже currentSongIndex. Но это относительно мелочи, важные но мелочи

g4rfan commented on main.cpp:
```diff
@@ -466,4 +213,126 @@ const char* vremenoeNazvanyePesny = pesny_VTope[i];
+	}
+}
+		// процедура для сортировки структур данных, часто повторяющийся код в Case3, 4, 11 и Case10, 12
+void sortingStruct(int currSong, int all, bool chng, bool prnt, bool inCase3411, bool inCase1012) {
```
Тут тоже самое по именнование chng - лучше всё таки change и print. Название функции тоже не очень, звучит так будто ты сортируешь структуры. Но не одних структур ты не принимаешь

g4rfan commented on main.cpp:
```diff
@@ -466,4 +213,126 @@ const char* vremenoeNazvanyePesny = pesny_VTope[i];
+}
+		// процедура для сортировки структур данных, часто повторяющийся код в Case3, 4, 11 и Case10, 12
+void sortingStruct(int currSong, int all, bool chng, bool prnt, bool inCase3411, bool inCase1012) {
+		std::string TMP = allSongs[currSong].track;
```
не знаю, считаешь ли ты что ты закончил задание, но вот тут явно не очень TMP и TMP2 по названию этих переменных ясно только одно - там временная переменная, могу лия яеё переопредеить и важна ли она - не ясно. Может это общая переменная для всего кода, или она всё таки важна. лучше перемименновать.

g4rfan commented on main.cpp:
```diff
@@ -466,4 +213,126 @@ const char* vremenoeNazvanyePesny = pesny_VTope[i];
+	}
+}
+
+void topSongsSorting(int all) { // сортировка песен по популярности, используемая в Case10 и 12
```
ты пишешь что здесь происходит сортировка, но по факту здесь модифицируются какие-то глобавльные массивы и никакой сортировки нет, ты вводишь в заблуждение тех кто будет пользоваться этой функцией

g4rfan commented:
В общем случае в этом решении есть несколько важных проблем, твой код очень сильно связан. И он даже связан не в привычном понятии связанности - ты завязался  не просто на какие-то модули или функции,  а на место вызова данной функци. Все эти inCase1012 inCase6 - это очень плохо, ты прям привязал код к конкретным case. А что если завтра кто-то захочет  добавить/удалить функцию или просто изменить порядок, все эти аргументы уже будут неактуальны и будут только запутывать, нужно стремиться к минимальной связанности кода.

Рассмотрим следующий момент, вот у тебя есть простая функция: void printInformation(int currSong, int all);  она выводит информацию о воспроизводимом треке, ок. Остаётся вопрос, а откуда она берёт эти треки? судя по сигнатуре функции туда прилетает текущая песня и все (но я об этом уже писал). Массив песен туда не летит, значит он как-то по другому узнаёт о песнях, например песни это глобальный массив. И вот это очень плохо, получается любая функция может поменять его случайно, и потом ты долго будешь искать кто же меняет массив лишний раз. Всё таки лучше чтобы массивы были локальны. 

Тут кстати всё ещё можно выделить структуры данных и обыграть некоторые функции по инному.

g4rfan commented on main.cpp:
```diff
@@ -466,4 +213,126 @@ const char* vremenoeNazvanyePesny = pesny_VTope[i];
+	needChange = false; needPrint = false; inSortCase = false; inShowCase = false;
+}
+	// процедура для поиска композиции по ее названию или названию группы
+void searchingSong(int all, std::string st, bool inCase5, bool inCase6) {
```
А зачем сюда вообще прилетает st, если через 3 строчки она затирается, а tempStr вообще не используется, и вообще что такое st, str и tempStr, что там должно лежать, вообще из названия не ясно

Kirchik replied:
Действительно, tempStr была лишней, а str - некорректное название переменной. Но в st передается строка, которая далее используется в процедуре для вывода. Зачем я передаю строку отвечу ниже...

g4rfan commented on main.cpp:
```diff
@@ -1,462 +1,209 @@
-    massiv_pesen[8] = "Heroes";
-    massivGrupp[8] = "David Bowie";
-    popularnost_pesen[8] = 0;
+bool needChange, needPrint, inSortCase, inShowCase, inBandNameCase, inSongNameCase;
```
Вот эти переменные тоже костыли. Например тут: 
`sortingStruct(numberOfTheCurrentSong, amountOfAllSongs, needChange, needPrint, inSortCase, inShowCase);`  - эта функция сортирует какие-то структуры, почему она что-то выводит в консоль, она должна только сортировать исходя из названия. needChange - что нужно поменять в сортировке, разве сортировка не работает по принципу изменения позиции элементов массива, что ещё надо менять и почему это зависит от какой-то глобальной перменной ?

Kirchik replied:
Вывод в консоль и замену номера песни убрал из данной процедуры. Насчет глобальной переменной -  в разных местах повторяется код за исключением лишь одного элемента - это может быть различие в строке: не "song", а "band", сортировка с другим условием: сортировка не по названию трека, а по популярности, хотя код везде одинаковый, за исключением лишь вот этого одного места. Для этого я сделал глобальные переменные тех строк, которые различаются и сделал проверку: "сейчас нужно произвести сортировку по названию или по популярности?", и эти глобальные переменные я передаю в функцию в месте ее вызова.

Kirchik committed:
Fix.

Вынес процедуру вывода информации из процедуры сортировки песен, сделал отдельную процедуру для изменения номера трека.
Изменил названия аргументов процедур.
Изменил названия некоторых процедур относительно их задач.
Изменил названия строковых переменных в процедурах.
Изменил названия булевых переменных, сократил их количество в аргументах процедур, теперь названия булевых переменных не привязаны к case и говорят какой код будет выполнен в зависимости от их значения.

Kirchik committed:
Передаю структуру в в функцию.

Передаю структуры данных allSongs и topList как аргументы в функцию чтобы обезопаситься от случайного изменения массива данных.

g4rfan commented on main.cpp:
```diff
@@ -466,4 +228,123 @@ const char* vremenoeNazvanyePesny = pesny_VTope[i];
+}
+
+	// процедура для поиска композиции по ее названию или названию группы
+void searchingSong(int amountOfSongs, std::string stringForOutputtingBandOrSong, Song Songs[]) {
```
на самом деле лучше было бы на данном этапе разделить вот такие функции на 2 разные функции, это добавляет читаемость и назвать например можно типа searchSongByName и searchSongByBand + уходят эти глобавльные переменные

g4rfan replied:
Также лучше чтобы функция поиска возвращала номер элемента где в массиве лежит данные элемент, а не сама решала что выводить консоль, эта функция только ищет, она не должна ещё и в консоль что-то выводить

g4rfan commented on main.cpp:
```diff
@@ -466,4 +228,123 @@ const char* vremenoeNazvanyePesny = pesny_VTope[i];
+}
+		// процедура для сортировки структур данных, часто повторяющийся код в Case3, 4, 11 и Case10, 12
+void sortingSongs(int currentSongIndex, int amountOfSongs, Song Songs[], Popular TopSongs[]) {
+	if (needToCompareTwoStrings) {	// условие для Case3, 4, 11
```
теже замечания что и ниже

g4rfan commented on main.cpp:
```diff
@@ -1,462 +1,224 @@
-
-                            std::cout << "8) Play specific song (specified by position) " << std::endl;
+struct Popular {
+	std::string topSongs;
```
Во-первых здесь не очень хорошие названия полей - выглядет так будто это массив названий топовых песен, хотя на самом деле это название одной песни. Также в английском музыкальная группа - это band, а просто группа как скопление чего-то это group, так что наверно лучше topBand. На а если свосем по существу - эта структура полностью повторяет по смыслу структуру Song, она просто не нужна, здесь всё ещё можно выделить другую структуру но не структуру Popular.

Kirchik committed:
Fix. Корректная сортировка популярности. Название переменных.

Заменил название переменной group на band.
Исправил баг, при котором сортировка песен по популярности не
срабатывала.
Удалил структуру Popularity, задал структуру topList через Song.

Kirchik committed:
Разделил функции.

Теперь вызов функций не зависит от глобальных переменных.

Kirchik committed:
Добавил файл заголовка.

Вынес функции в отдельный модуль.

Kirchik committed:
Разделил файлы интерфейса и файлы реализации.
